### DISCUSSION TITLE
When/How do api route files recompile?

### DISCUSSION THREAD

**Discussion by neeraj3029** [vercel/next.js#36032] [Help]
I am using the an API route `/api/metrics` to expose an endpoint that shares the metrics of node server.

The problem occurs when the the `/api/metrics.ts` file recompiles, causing the lib to initialised again (which causes other errors).

Is there any way I can ensure that the route file is compiled only once?

**Comment by icyJoseph** [comment]
Hi,

What do you mean with `"recompiles"`?

API routes are serverless functions, which are spin-up and down, as needed. This is done in order to consume as little resources as possible and prevent memory leaks. You should not assume that API routes, are the same instance through time. 

See this video for a more versatile explanation: https://youtu.be/jv4Soz08gBw?t=1315

I am not sure if you can tell the next server to try and keep these alive for as long as possible, but even if you could, if you deploy to Vercel for example, they'll timeout and be shutdown. I think the same applies for AWS amplify and others.

I think you may need to look into how to externally resolve the api routes. I haven't done this myself, but at the very least the API route needs to export configuration, like so, [from the docs](https://nextjs.org/docs/api-routes/api-middlewares#custom-config):

```ts
export const config = {
  api: {
    externalResolver: true,
  },
}
``` 

Here's an [example](https://github.com/vercel/next.js/tree/main/examples/custom-server-express) using `express`.

I'd suggest trying to design your application so that these metrics are not stateful, or could be stored elsewhere, and do not depend on ever running serverless functions.
